core/src/main/java/org/bitcoinj/core/TransactionOutPoint.java

```diff
@@ -139,6 +189,36 @@ public byte[] bitcoinSerialize() {
         public int getMessageSize() {
             return BYTES;
         }
+    }
+
+    class Connected extends Unconnected implements TransactionOutPoint  {
+
+        /** Special outpoint that normally marks a coinbase input. It's also used as a test dummy. */
+        public static final Connected UNCONNECTED =
+                new Connected(ByteUtils.MAX_UNSIGNED_INTEGER, Sha256Hash.ZERO_HASH);
+
+
+        // This is not part of bitcoin serialization. It points to the connected transaction.
+        Transaction fromTx;
+
+        // The connected output.
+        TransactionOutput connectedOutput;
+
+        public Connected(long index, Transaction fromTx) {
+            super(fromTx.getTxId(), index);
+            this.fromTx = fromTx;
+        }
+
+        public Connected(long index, Sha256Hash hash) {
+            super(hash, index);
+            checkArgument(index >= 0 && index <= ByteUtils.MAX_UNSIGNED_INTEGER, () ->
+                    "index out of range: " + index);
+        }
```

Comment:
Why do these exist? Seems like they only create an unconnected output.

Perhaps we should add an assert to the `Connected` constructors: at least one of `fromTx` and `connectedOutput` must be set. Or even `fromTx` always, I'm not sure. But both `null` should be disallowed at this point, otherwise we wouldn't have the `Unconnected` subclass.

Reply:
Because currently the `Connected` class is a "transformer" it can "connect" an "unconnect".  So I've only updated code to use `Unconnected` when those instances never need to transform. Further steps of refactoring will be needed to handle connection and unconnection properly.